src/model_selection/kFold.ts

```diff
@@ -0,0 +1,161 @@
+
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { tf } from '../shared/globals'
+import { Tensor2D, Tensor1D } from '@tensorflow/tfjs-node'
+import { assert } from '../typesUtils'
+import { CrossValidator } from './crossValidator'
+import { alea } from 'seedrandom'
+import * as randUtils from '../randUtils'
+
+export interface KFoldParams {
+  /**
+   * Number of ways in which the dataset is to be split.
+   * Defaults to `5`. Larger numbers of splits will result
+   * in increased computational cost, since the model is
+   * trained `nSplits` time during cross validation. In
+   * return, more training data is available in each split,
+   * which allows utilize smaller datasets more efficiently.
+   */
+  nSplits?: number
+  /**
+   * If set to `true`, indices are shuffled before train/test
+   * splitting. Defaults to `false`.
+   */
+  shuffle?: boolean
+  /**
+   * Random seed to be used for shuffling. Ignored if `nSplits = false`.
+   */
+  randomState?: number
+}
+
+/**
+ * K-Fold cross-validator
+ *
+ * Generates train and test indices to split data in train/test subsets.
+ * To generate these subsets, the dataset is split into k (about) evenly
+ * sized chunks of consecutive elements. Each split takes another chunk
+ * as test data and the remaining chunks are combined to be the training
+ * data.
+ *
+ * Optionally, the indices can be shuffled before splitting it into chunks
+ * (disabled by default).
+ *
+ * @example
+ * ```js
+ * import { KFold } from 'scikitjs'
+ *
+ * const kf = new KFold({ nSplits: 3 })
+ *
+ * const X = tf.range(0, 7).reshape([7, 1]) as Tensor2D
+ *
+ * console.log( 'nSplits:', kf.getNumSplits(X) )
+ *
+ * for (const { trainIndex, testIndex } of kf.split(X) )
+ * {
+ *   try {
+ *     console.log( 'train:', trainIndex.toString() )
+ *     console.log( 'test:',   testIndex.toString() )
+ *   }
+ *   finally {
+ *     trainIndex.dispose()
+ *      testIndex.dispose()
+ *   }
+ * }
+ * ```
+ */
+export class KFold implements CrossValidator {
+  nSplits: number
+  shuffle: boolean
+  randomState?: number
+
+  constructor({
+    nSplits = 5,
+    shuffle = false,
+    randomState
+  }: KFoldParams = {}) {
+    nSplits = Number(nSplits)
+    assert(
+      nSplits % 1 === 0 && nSplits > 1,
+      'new KFold({nSplits}): nSplits must be an int greater than 1.'
+    )
+    this.nSplits = nSplits
+    this.shuffle = !!shuffle
+    this.randomState = randomState
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public getNumSplits(X?: Tensor2D, y?: Tensor1D, groups?: Tensor1D): number {
+    return this.nSplits
+  }
+
+  public *split(
```

Comment:
A suggestion: shouldn't we support other data types and not just Tensors? I believe the Scikit2D and Scikit1D types covers those